### PR TITLE
[Draft] WIP: Add eventing command to run Knative Eventing scaling tests (first version MVP)

### PR DESCRIPTION
**Pull request by aslom**
This PR implements kperf eventing subcommand following design discussed in https://github.com/knative-sandbox/kperf/issues/24

Getting close to MVP. Sharing Work-in-progress including docs on how to run current version is described in eventing.md (development notes about building Docker image and local testing are in eventing-dev.md)

Remaining work for MVP:
[ ] prepare: support multiple namespaces 
[ ] measure: factor out and improve event workload generator
[ ] measure: add latency buckets and measuring errors, duplicates
[ ] measure: add printing output to .csv, option to stop measurement command when no changes in metrics after N seconds 
[ ] measure: add missing functionality in one command - currently split outside: start kubectl tunnel(s) to receiver namespaces, start retrieving metrics from tunneled localhost:8001 (8002 ...), deploy driver job 
(alternatively instead of tunnels in background https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-services/#manually-constructing-apiserver-proxy-urls )
[ ] cleanup: for each namespace remove driver jobs and receiver deloyment, pods
[ ] add unit tests
[ ] fix build issues (mod vendoring?)
[ ] fix Kafka receiver blocking for Ctrl-C

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# Retrieve application dependencies using go modules.
+# Allows container builds to reuse downloaded dependencies.
+COPY go.* ./
+RUN go mod download
+
+# Copy local code to the container image.
+COPY . ./
+
+
+# TODO use hack/build.sh ?
```
Comment (maximilien) [inline Dockerfile:17]: Yes, otherwise remove this TODO

File: eventing.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
+# kperf eventing
+
+## Temporary setup
+
+That functionality will be wrapped into `kperf eventing` command in near future.
```
Comment (lionelvillard) [inline eventing.md:5]: isn't this already the case?

File: eventing.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+See separate docs about [kpef eventing development](./eventing-dev.mg).
+
+## Running performance tests
+
+### kperf prepare
+
+That functionality will be wrapped into `kperf eventing prepare` command in near future.
+
+Currently it needs ti be run from command line to dwpeloy test receiver.
```
Comment (lionelvillard) [inline eventing.md:28]: nit: dwpeloy -> deploy

File: eventing.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+```
+go run -mod=readonly ./cmd/ eventing measure
+```
+
+
+#### Start test drviver
+
+
+Determina location of receiver service in Kubernetes namespace:
```
Comment (lionelvillard) [inline eventing.md:87]: nit: Determina -> Determine

File: pkg/command/eventing/clean.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Copyright 2020 The Knative Authors
```
Comment (lionelvillard) [inline pkg/command/eventing/clean.go:1]: 2021?

File: pkg/command/eventing/clean.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+
+	"github.com/spf13/cobra"
+
+	"knative.dev/kperf/pkg"
+)
+
+func NewEventingCleanCommand(p *pkg.PerfParams) *cobra.Command {
+	cleanArgs := cleanArgs{}
+	ksvcCleanCommand := &cobra.Command{
```
Comment (lionelvillard) [inline pkg/command/eventing/clean.go:29]: better use eventingCleanCommand

File: pkg/command/eventing/clean.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		Use:   "clean",
+		Short: "Clean eventing workload",
+		Long: `Clean eventing workload
+
+For example:
+# To clean Knative Eventing workload
+kperf eventing clean --namespace-prefix testns / --namespace nsname
```
Comment (lionelvillard) [inline pkg/command/eventing/clean.go:36]: `/` looks strange here.

File: pkg/command/eventing/clean.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			fmt.Println("Eventng cleaning called")
```
Comment (lionelvillard) [inline pkg/command/eventing/clean.go:40]: nit: `Eventng`-> `Eventing`

File: pkg/command/eventing/clean.go (hunk 5 above)
Comment (lionelvillard) [reply pkg/command/eventing/clean.go:1]: this is actually all over the place

File: pkg/command/eventing/driver.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventing
+
+import (
+	"fmt"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
```
Comment (lionelvillard) [inline pkg/command/eventing/driver.go:20]: no need here. Only main should include it.

File: pkg/command/eventing/driver.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/spf13/cobra"
+
+	"knative.dev/kperf/pkg"
+	"knative.dev/kperf/pkg/command/eventing/driver"
+)
+
+func NewEventingDriverCommand(p *pkg.PerfParams) *cobra.Command {
+	cleanArgs := cleanArgs{}
+	ksvcCleanCommand := &cobra.Command{
```
Comment (lionelvillard) [inline pkg/command/eventing/driver.go:30]: ksvcCleanCommand -> `eventingCleanCommand`

File: pkg/command/eventing/prepare.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+import (
+	"errors"
+	"html/template"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
```
Comment (lionelvillard) [inline pkg/command/eventing/prepare.go:25]: remove

**Comment by lionelvillard** [comment]
Added some comments. I made sure it does not break Serving. 

LGTM

File: pkg/command/eventing/driver/kafka_sender.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+	//Port int `envconfig:"PORT" default:"8080"`
+
+	// KafkaServer URL to connect to the Kafka server.
+	KafkaServer string `envconfig:"KAFKA_BOOTSTRAP_SERVERS" required:"true"`
+
+	// Subject is the nats subject to publish cloudevents on.
+	Topic string `envconfig:"KAFKA_TOPIC" required:"true"`
+}
+
+func sendEvents(plan SendEventsPlan) EventsStats {
```
Comment (maximilien) [inline pkg/command/eventing/driver/kafka_sender.go:62]: Break this down into smaller helper function and that will allow you to improve and better visualize this code compared to the Redis sender and others. Also, makes it easier to debug and review :)

File: pkg/command/eventing/driver/null_sender.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+type NullEventSender struct {
+	Plan             SendEventsPlan
+	SleepTimeSeconds float64 // simulate sending time
+}
+
+func (s NullEventSender) Send() EventsStats {
+	plan := s.Plan
+	eventsSent := int(float64(plan.eventsPerSecond) * plan.durationSeconds)
+	// TODO better golang idiom/conversion?
```
Comment (maximilien) [inline pkg/command/eventing/driver/null_sender.go:27]: I think this is fine and idiomatic. Remove TODO

File: pkg/command/eventing/driver/redis_sender.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+	"github.com/gomodule/redigo/redis"
+)
+
+type RedisEventSender struct {
+	Plan    SendEventsPlan
+	Address string
+	// NumConsumers   string
+	TLSCertificate string
+	Stream         string
+	// Group          string
```
Comment (maximilien) [inline pkg/command/eventing/driver/redis_sender.go:41]: Remove commented fields

File: pkg/command/eventing/driver/redis_sender.go [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+			Type:   "com.example.FullEvent",
+			Source: source,
+			ID:     "full-event",
+			Time:   &timestamp,
+			//DataSchema: &schema,
+			Subject: strptr("topic"),
+		}.AsV1(),
+	}
+	//if err := event.SetData("text/json", "[\"fruit\", \"orange\"]"); err != nil {
+	data := []byte("[\"fruit\", \"orange\"]")
```
Comment (maximilien) [inline pkg/command/eventing/driver/redis_sender.go:81]: This seems like test data...

File: pkg/command/eventing/driver/redis_sender.go [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+	//data := []byte("{\"a\" : \"b\"}") // {"a": "b"}
+	if err := e.SetData(event.ApplicationJSON, data); err != nil {
+		panic(err)
```
Comment (maximilien) [inline pkg/command/eventing/driver/redis_sender.go:84]: Same as before these panic and logFatal will come back to hunt you when you need to debug. Best just deal with error at first catch with message, return and cascade up to end cleanly

File: pkg/command/eventing/driver/workload_generator.go [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+func (s EventGenerator) NextCloudEvents() []*event.Event {
+	e := event.Event{}
+	events := make([]*event.Event, 1, 100)
+	events[0] = &e
+	return events
+}
+
+func (s EventGenerator) NextCloudEventsAsMaps() []*map[string]string {
+	events := make([]*map[string]string, 1, 100)
+	values := map[string]string{"id": "1234668888888", "source": "323223232332909090", "type": "dev.knative.eventing.test.scaling", "timestamp": "12929299999992222"}
```
Comment (maximilien) [inline pkg/command/eventing/driver/workload_generator.go:41]: Why do we have test data in the main code? Shouldn’t these be loaded from config if you need to have some values?

File: pkg/command/eventing/driver_test.go [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+// limitations under the License.
+
+package eventing
+
+import (
+	"testing"
+)
+
+func TestNewEventingDriverCommand(t *testing.T) {
+
```
Comment (maximilien) [inline pkg/command/eventing/driver_test.go:22]: Same as before. What I would do is to create an issue with a list of all the missing tests.

File: pkg/command/eventing/eventing.go [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/spf13/cobra"
+	"knative.dev/kperf/pkg"
+)
+
+// domainCmd represents the domain command
+func NewEventingCmd(p *pkg.PerfParams) *cobra.Command {
+	var eventingCmd = &cobra.Command{
+		Use:   "eventing",
+		Short: "Knative eventing load test",
+		Long: `Knative eventing load test and measurement. For example:
```
Comment (maximilien) [inline pkg/command/eventing/eventing.go:27]: Use # comment to separate what the user sees as comment vs code they can cut and paste to try

File: pkg/command/eventing/eventing_test.go [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+	"testing"
+
+	"gotest.tools/v3/assert"
+)
+
+func TestNewEventingCmd(t *testing.T) {
+	cmd := NewEventingCmd(nil)
+	assert.Check(t, cmd.HasSubCommands(), "cmd eventing should have subcommands")
+
+	// _, _, err := cmd.Find([]string{"prepare"})
```
Comment (maximilien) [inline pkg/command/eventing/eventing_test.go:27]: Same as before, remove these empty tests and create issue with a TODO list.

If you want to keep this code then move it to a branch. Use github for what’s it good for, organizing your code :)

**Review by maximilien — CHANGES_REQUESTED** [review]
Lots and lots of issues. Some small, nits, and larger. Seems like many of the issues have to do with imcomplete code. So perhaps this is not really an MVP but rather an incomplete v1.0.

I would love this to move fast. Perhaps one approach is to submit a real MVP — that is only the parts that are done and create issues for the rest. For instance, only support Kafka and move the rest as issues to be submitted in subsequent PRs.

Also, instead of having a bunch of empty unit tests move these to a issue to be done as future PR. And focus on having e2e tests only for MVP?

Finally, I want to move fast in my reviews, however, long PR mean longer time to review (obvious ;) so perhaps many small PRs (even if they depend on each other is a better approach to move faster?

All that said, many thanks @aslom for getting this done and for your hard work. Truly appreciated. Cheers 🍻

**Comment by maximilien** [comment]
Any update on this PR @aslom ?

**Comment by aslom** [comment]
@maximilien working on code simplification and refactoring - should be ready in few days

**Comment by maximilien** [comment]
Hi, @aslom any updates on this. Anything I can do to help move this?

**Comment by maximilien** [comment]
Hi, @aslom any updates on this?

**Comment by aslom** [comment]
@maximilien I have done lot of testing and improvements and finding kperf eventing very useful in testing - I have changed to PR titleWIP to do indicate there will be restructuring going on

**Comment by maximilien** [comment]
Hi, @aslom, do we have an update on this?

**Pull request closed without merging**